cppapi/client/group.cpp

```diff
@@ -48,63 +49,109 @@ bool GroupReply::exception_enabled = false;
+    const std::string& name_or_pattern,
+    int timeout_millis)
 {
-#if defined(_LOCAL_DEBUGGING)
-  cout << "GroupElementFactory::instanciate::pattern [" << p << "]" << std::endl;
-#endif
-  //- a vector to store device names
-  std::vector<std::string> dnl(0);
-  //- is <p> a device name or a device name pattern ?
-  if (p.find('*', 0) == std::string::npos)
-  {
-#if defined(_LOCAL_DEBUGGING)
-    cout << "\t|- pattern is pure device name" << std::endl;
-#endif
-    dnl.push_back(p);
-  }
-  else
-  {
- 	  int db_port = 0;
-    std::string db_host, new_pattern;
-	  DbDatum dbd;
-	  parse_name(p, db_host, db_port, new_pattern);
-
-    if (db_host.size() == 0) {
-  	  Database db;
-      //- ask the db the list of device matching pattern p
-  	  dbd = db.get_device_exported(const_cast<std::string&>(p));
-    }
-	  else {
-		  ApiUtil *au = ApiUtil::instance();
-		  int db_ind = au->get_db_ind(db_host,db_port);
-		  dbd = ((au->get_db_vect())[db_ind])->get_device_exported(const_cast<std::string&>(new_pattern));
-	  }
-
-    //- extract device names from dbd
-    dbd >> dnl;
-#if defined(_LOCAL_DEBUGGING)
-    cout << "\t|- db.get_device_exported::found ";
-    cout << dnl.size() << " device names matching pattern" << std::endl;
-    for (unsigned int dn = 0; dn < dnl.size(); dn++) {
-      cout << "\t\t|- " << dnl[dn] << std::endl;
+    return create_group_elements(resolve_device_names(name_or_pattern), timeout_millis);
+}
+
+GroupElements GroupElementFactory::create_group_elements(const DeviceNames& names, int timeout_millis)
+{
+    GroupElements group_elements;
+    group_elements.reserve(names.size());
+
+    for (DeviceNames::const_iterator name = names.begin(); name != names.end(); ++name)
+    {
+        group_elements.push_back(new GroupDeviceElement(*name, timeout_millis));
     }
-#endif
-  }
-  //- build the returned GroupElementList
-  GroupElements tel(0);
-  GroupDeviceElement* tde;
-  for (unsigned int i = 0; i < dnl.size(); i++) {
-    tde = new GroupDeviceElement(dnl[i],tmo_ms);
-    if (tde) {
-      tel.push_back(tde);
+
+    return group_elements;
+}
+
+DeviceNames GroupElementFactory::resolve_device_names(const std::string& name_or_pattern)
+{
+    const bool is_name = name_or_pattern.find('*') == std::string::npos;
+
+    if (is_name)
+    {
+        return DeviceNames(1, name_or_pattern);
     }
-  }
-#if defined(_LOCAL_DEBUGGING)
-    cout << "\t|- GroupElementList contains " << tel.size() << " elements" << std::endl;
-#endif
-  //- return the list to the caller
-  return tel;
+    else
+    {
+        int db_port = 0;
+        std::string db_host = "";
+        std::string name_or_pattern_without_host = "";
+        parse_name(name_or_pattern, db_host, db_port, name_or_pattern_without_host);
+
+        const bool is_local_device = db_host.empty();
+
+        if (is_local_device)
+        {
+            return resolve_local_device_names(name_or_pattern);
+        }
+        else
```

Comment:
This will look like a "guard" expression then:
```cpp
if (some_precondition)
{
    return handle_special_case()
}
work_as_usual()
```
If both branches are equally important and equally likely to happen I like to keep the `else`. See https://www.fluentcpp.com/2018/02/20/on-using-guards-in-cpp/, especially "Don’t sacrifice symmetry to reduce nesting" section for a deeper explanation.